File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/mapper/BookMapper.java
package ru.elenacherev.librarymanager.mapper;

import ru.elenacherev.librarymanager.api.dto.Book;
import ru.elenacherev.librarymanager.domain.entity.BookEntity;
import ru.elenacherev.librarymanager.domain.entity.EditionEntity;

public class BookMapper {
    public static Book map(BookEntity entity) {
        return Book.builder()
                .bookId(entity.getBookId())
                .isAvailable(entity.isAvailable())
                .isReserved(entity.isReserved())
                .shelfNumber(entity.getShelfNumber())
                .title(entity.getEdition()!=null
                		?entity.getEdition().getTitle()
                		: ""
                      )
                .editionId(entity.getBookId())
                .build();
    }

    public static BookEntity map(Book dto, EditionEntity edition) {
        return map(dto, new BookEntity(), edition);
    }

    public static BookEntity map(
    		Book dto, 
    		BookEntity entity,
    		EditionEntity edition
    ) {
        entity.setAvailable(dto.isAvailable());
        entity.setReserved(dto.isReserved());
        entity.setShelfNumber(dto.getShelfNumber());
        entity.setEdition(edition);
        return entity;
    }
}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/mapper/BookMapper.kt
package ru.elenacherev.librarymanager.mapper

import ru.elenacherev.librarymanager.api.dto.Book
import ru.elenacherev.librarymanager.domain.entity.BookEntity
import ru.elenacherev.librarymanager.domain.entity.EditionEntity

fun BookEntity.map() = Book(
    bookId = bookId,
    isAvailable = isAvailable,
    isReserved = isReserved,
    shelfNumber = shelfNumber,
    editionId = edition.editionId!!,
    title = edition.title
)

fun Book.map(edition: EditionEntity) = BookEntity(
    isAvailable = isAvailable,
    isReserved = isReserved,
    shelfNumber = shelfNumber,
    edition = edition
)

fun Book.map(
    entity: BookEntity,
    edition: EditionEntity
) = entity.also {
    it.isAvailable = isAvailable
    it.isReserved = isReserved
    it.shelfNumber = shelfNumber
    it.edition = edition
}